// @mintlify/mdx/dist/client/rsc.js
import { MDXRemote as BaseMDXRemote } from 'next-mdx-remote-client/rsc';
import { jsx as _jsx } from 'react/jsx-runtime';
import rehypeKatex from 'rehype-katex';
import remarkGfm from 'remark-gfm';
import remarkMath from 'remark-math';
import remarkSmartypants from 'remark-smartypants';

import { rehypeSyntaxHighlighting } from '../plugins/index.js';
import { Popup, PopupContent, PopupTrigger } from '../ui/index.js';

export async function MDXRemote({
  source,
  mdxOptions,
  scope,
  components,
  parseFrontmatter,
  syntaxHighlightingOptions,
}) {
  const mergedComponents = {
    Popup,
    PopupContent,
    PopupTrigger,
    ...components,
  };
  return (
    // @ts-expect-error Server Component
    _jsx(BaseMDXRemote, {
      source: source,
      components: mergedComponents,
      options: {
        scope,
        mdxOptions: {
          remarkPlugins: [
            remarkGfm,
            remarkSmartypants,
            remarkMath,
            ...(mdxOptions?.remarkPlugins || []),
          ],
          rehypePlugins: [
            rehypeKatex,
            [rehypeSyntaxHighlighting, syntaxHighlightingOptions],
            ...(mdxOptions?.rehypePlugins || []),
          ],
          format: mdxOptions?.format || 'mdx',
        },
        parseFrontmatter,
      },
    })
  );
}
