// @mintlify/validation/dist/mint-config/schemas/v2/properties/reusable/page.js
import { apiPlaygroundDisplayTypes, iconTypes } from '@mintlify/models';
import { z } from 'zod';
import { normalizeRelativePath } from '../../../../transforms/normalizeRelativePath.js';
import { iconSchema } from './icon.js';
const relatedFrontmatterEntrySchema = z.union([
    z.string(),
    z.record(z.string(), z.string()).refine((entry) => Object.keys(entry).length === 1, {
        message: 'Related page title entries must contain exactly one link',
    }),
]);
const relatedFrontmatterSchema = z.union([
    z.literal(false),
    z.array(relatedFrontmatterEntrySchema),
]);
export const pageSchema = z
    .string()
    .nonempty()
    .transform(normalizeRelativePath)
    .describe('A page in the navigation. Referenced by the path to the page. Example: path/to/page');
export const decoratedPageSchema = z
    .object({
    href: z.string(),
    title: z.string(),
    sidebarTitle: z.string().optional(),
    description: z.string().optional(),
    canonical: z.string().optional(),
    'og:title': z.string().optional(),
    'og:url': z.string().optional(),
    api: z.string().optional(),
    openapi: z.string().optional(),
    asyncapi: z.string().optional(),
    contentType: z.string().optional(),
    authMethod: z.string().optional(),
    auth: z.string().optional(),
    version: z.string().optional(),
    mode: z.string().optional(),
    hideFooterPagination: z.boolean().optional(),
    authors: z.unknown().optional(),
    lastUpdatedDate: z.string().optional(),
    createdDate: z.string().optional(),
    'openapi-schema': z.string().optional(),
    icon: iconSchema.optional(),
    iconType: z.enum(iconTypes).optional().describe('Deprecated, use icon instead'),
    tag: z.string().optional(),
    url: z.string().optional(),
    hideApiMarker: z.boolean().optional(),
    noindex: z.boolean().optional(),
    isPublic: z.boolean().optional(),
    public: z.boolean().optional(),
    boost: z.number().optional(),
    deprecated: z.boolean().optional(),
    hidden: z.boolean().optional(),
    groups: z.array(z.string()).optional(),
    robots: z.string().optional(),
    playground: z.enum(apiPlaygroundDisplayTypes).optional(),
    keywords: z.array(z.string()).optional(),
    timestamp: z.boolean().optional(),
    related: relatedFrontmatterSchema.optional(),
})
    .passthrough()
    .describe('page metadata tags in the navigation');
