// @mintlify/validation/dist/resolve-refs.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
export class RefNotFoundError extends Error {
    constructor(refPath, referencedFrom) {
        super(`Referenced config file '${refPath}' does not exist (referenced from ${referencedFrom})`);
        this.name = 'RefNotFoundError';
    }
}
export class CircularRefError extends Error {
    constructor(chain) {
        super(`Circular reference detected: ${chain.join(' → ')}`);
        this.name = 'CircularRefError';
    }
}
export class PathTraversalError extends Error {
    constructor(refPath) {
        super(`Referenced config file '${refPath}' is outside the project directory`);
        this.name = 'PathTraversalError';
    }
}
export class RefInvalidJsonError extends Error {
    constructor(refPath, details) {
        super(`Referenced config file '${refPath}' contains invalid JSON: ${details}`);
        this.name = 'RefInvalidJsonError';
    }
}
export function isJsonObject(value) {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
}
export function getRefPath(value) {
    if (!isJsonObject(value) || typeof value.$ref !== 'string')
        return undefined;
    return value.$ref;
}
export function normalizePath(path) {
    const normalizedPath = path.replace(/\\/g, '/');
    const isAbsolute = normalizedPath.startsWith('/');
    const segments = normalizedPath.split('/');
    const normalizedSegments = [];
    for (const segment of segments) {
        if (segment === '' || segment === '.')
            continue;
        if (segment === '..') {
            const lastSegment = normalizedSegments.at(-1);
            if (!lastSegment || lastSegment === '..') {
                normalizedSegments.push(segment);
            }
            else {
                normalizedSegments.pop();
            }
            continue;
        }
        normalizedSegments.push(segment);
    }
    const joined = normalizedSegments.join('/');
    return isAbsolute ? `/${joined}` : joined;
}
export function dirname(filePath) {
    const normalizedPath = normalizePath(filePath);
    const lastSlashIndex = normalizedPath.lastIndexOf('/');
    if (lastSlashIndex === -1)
        return '';
    return normalizedPath.slice(0, lastSlashIndex);
}
export function resolveRefPath(refPath, fromFile) {
    if (!fromFile)
        return normalizePath(refPath);
    return normalizePath([dirname(fromFile), refPath].filter(Boolean).join('/'));
}
function isPathOutsideProjectRoot(filePath) {
    return filePath === '..' || filePath.startsWith('../') || filePath.startsWith('/');
}
function loadRefFile(refPath, fullPath, readFile, fromFile) {
    return __awaiter(this, void 0, void 0, function* () {
        let content;
        try {
            content = yield readFile(fullPath);
        }
        catch (_a) {
            throw new RefNotFoundError(refPath, fromFile !== null && fromFile !== void 0 ? fromFile : 'project root');
        }
        try {
            return { parsed: JSON.parse(content), content };
        }
        catch (error) {
            throw new RefInvalidJsonError(refPath, error instanceof Error ? error.message : String(error));
        }
    });
}
export function resolveRefs(value, readFile) {
    return __awaiter(this, void 0, void 0, function* () {
        const referencedFiles = [];
        const referencedFileContents = {};
        function resolve(currentValue, visited, fromFile) {
            return __awaiter(this, void 0, void 0, function* () {
                const refPath = getRefPath(currentValue);
                if (refPath != null) {
                    const fullPath = resolveRefPath(refPath, fromFile);
                    if (isPathOutsideProjectRoot(fullPath)) {
                        throw new PathTraversalError(refPath);
                    }
                    if (visited.has(fullPath)) {
                        throw new CircularRefError([...visited, fullPath]);
                    }
                    const { parsed, content } = yield loadRefFile(refPath, fullPath, readFile, fromFile);
                    referencedFiles.push(fullPath);
                    referencedFileContents[fullPath] = content;
                    const resolvedRef = yield resolve(parsed, new Set([...visited, fullPath]), fullPath);
                    if (isJsonObject(resolvedRef) && isJsonObject(currentValue)) {
                        const siblingEntries = Object.entries(currentValue).filter(([key]) => key !== '$ref');
                        const result = Object.assign({}, resolvedRef);
                        for (const [key, siblingValue] of siblingEntries) {
                            result[key] = yield resolve(siblingValue, visited, fromFile);
                        }
                        return result;
                    }
                    return resolvedRef;
                }
                if (Array.isArray(currentValue)) {
                    return Promise.all(currentValue.map((item) => resolve(item, visited, fromFile)));
                }
                if (isJsonObject(currentValue)) {
                    const result = {};
                    for (const [key, childValue] of Object.entries(currentValue)) {
                        result[key] = yield resolve(childValue, visited, fromFile);
                    }
                    return result;
                }
                return currentValue;
            });
        }
        const resolved = yield resolve(value, new Set());
        return { resolved, referencedFiles, referencedFileContents };
    });
}
