// @mintlify/validation/dist/mint-config/schemas/v2/properties/navigation/groups.js
import { z } from 'zod';
import { asyncApiSchema } from '../reusable/asyncapi.js';
import { hiddenSchema } from '../reusable/hidden.js';
import { iconSchema } from '../reusable/icon.js';
import { openApiSchema } from '../reusable/openapi.js';
import { decoratedPageSchema, pageSchema } from '../reusable/page.js';
import { decoratedPagesSchema, pagesSchema } from './pages.js';
export const baseGroupSchema = z.object({
    group: z.string().nonempty().describe('The name of the group'),
    public: z.boolean().optional().describe('Whether the group is public'),
    icon: iconSchema.optional(),
    hidden: hiddenSchema.optional(),
    root: pageSchema.optional(),
    tag: z.string().optional().describe('Tag for the group'),
    expanded: z.boolean().optional().describe('Whether the group is expanded by default'),
});
export const groupSchema = z
    .union([
    baseGroupSchema.extend({
        openapi: openApiSchema,
        pages: z.lazy(() => pagesSchema).default([]),
    }),
    baseGroupSchema.extend({
        asyncapi: asyncApiSchema,
        pages: z.lazy(() => pagesSchema).default([]),
    }),
    baseGroupSchema.extend({ pages: z.lazy(() => pagesSchema).default([]) }),
])
    .describe('Organizing by groups');
export const decoratedGroupSchema = z
    .object({
    group: z.string().nonempty().describe('The name of the group'),
    public: z.boolean().optional().describe('Whether the group is public'),
    icon: iconSchema.optional(),
    hidden: hiddenSchema.optional(),
    root: decoratedPageSchema.optional(),
    pages: z.lazy(() => decoratedPagesSchema),
    tag: z.string().optional().describe('Tag for the group'),
    expanded: z.boolean().optional().describe('Whether the group is expanded by default'),
})
    .describe('Organizing by groups');
export const groupsSchema = z
    .array(groupSchema)
    .min(1, 'At least one group must be specified')
    .describe('Organizing by groups');
export const decoratedGroupsSchema = z
    .array(decoratedGroupSchema)
    .min(1, 'At least one group must be specified')
    .describe('Organizing by groups');
