// @mintlify/models/dist/types/assistantTier.js
export const TierIds = [
    'tier-0-250',
    'tier-251-750',
    'tier-751-1500',
    'tier-1501-3000',
    'tier-3001-5000',
    'tier-5001-7500',
    'tier-7501-10000',
    'tier-10001-15000',
    'tier-15001-20000',
    'tier-20001-25000',
    'tier-25001-35000',
    'tier-35001-50000',
    'tier-50001-plus',
];
export const ASSISTANT_TIER_OVERAGE_PRICE_PER_CREDIT = 0.25;
export const ASSISTANT_TIERS = {
    'tier-0-250': {
        name: '0-250',
        min: 0,
        max: 250,
        monthlyPrice: 0,
    },
    'tier-251-750': {
        name: '251-750',
        min: 251,
        max: 750,
        monthlyPrice: 125,
    },
    'tier-751-1500': {
        name: '751-1,500',
        min: 751,
        max: 1500,
        monthlyPrice: 300,
    },
    'tier-1501-3000': {
        name: '1,501-3,000',
        min: 1501,
        max: 3000,
        monthlyPrice: 633,
    },
    'tier-3001-5000': {
        name: '3,001-5,000',
        min: 3001,
        max: 5000,
        monthlyPrice: 1045,
    },
    'tier-5001-7500': {
        name: '5,001-7,500',
        min: 5001,
        max: 7500,
        monthlyPrice: 1523,
    },
    'tier-7501-10000': {
        name: '7,501-10,000',
        min: 7501,
        max: 10000,
        monthlyPrice: 1950,
    },
    'tier-10001-15000': {
        name: '10,001-15,000',
        min: 10001,
        max: 15000,
        monthlyPrice: 2802,
    },
    'tier-15001-20000': {
        name: '15,001-20,000',
        min: 15001,
        max: 20000,
        monthlyPrice: 3555,
    },
    'tier-20001-25000': {
        name: '20,001-25,000',
        min: 20001,
        max: 25000,
        monthlyPrice: 4207,
    },
    'tier-25001-35000': {
        name: '25,001-35,000',
        min: 25001,
        max: 35000,
        monthlyPrice: 5560,
    },
    'tier-35001-50000': {
        name: '35,001-50,000',
        min: 35001,
        max: 50000,
        monthlyPrice: 7462,
    },
    'tier-50001-plus': {
        name: '50,001+',
        min: 50001,
        max: 75000,
        monthlyPrice: 10465,
    },
};
