// @mintlify/models/dist/types/chat-attachment.js
export const IMAGE_EXTENSION_MIME_MAP = {
    jpg: 'image/jpeg',
    jpeg: 'image/jpeg',
    png: 'image/png',
    gif: 'image/gif',
    webp: 'image/webp',
    svg: 'image/svg+xml',
};
export const DOCUMENT_EXTENSION_MIME_MAP = {
    pdf: 'application/pdf',
};
export const CODE_EXTENSION_MIME_MAP = {
    js: 'text/javascript',
    jsx: 'text/javascript',
    ts: 'text/typescript',
    tsx: 'text/typescript',
    mjs: 'text/javascript',
    cjs: 'text/javascript',
    md: 'text/markdown',
    mdx: 'text/markdown',
    json: 'application/json',
    html: 'text/html',
    css: 'text/css',
    py: 'text/x-python',
    csv: 'text/csv',
    txt: 'text/plain',
    yaml: 'text/yaml',
    yml: 'text/yaml',
    xml: 'text/xml',
    sql: 'text/x-sql',
    sh: 'text/x-shellscript',
    bash: 'text/x-shellscript',
    zsh: 'text/x-shellscript',
    graphql: 'text/x-graphql',
    gql: 'text/x-graphql',
    toml: 'text/toml',
    env: 'text/plain',
    go: 'text/x-go',
    rs: 'text/x-rust',
    rb: 'text/x-ruby',
    java: 'text/x-java',
    kt: 'text/x-kotlin',
    swift: 'text/x-swift',
    c: 'text/x-c',
    cpp: 'text/x-c++',
    cc: 'text/x-c++',
    h: 'text/x-c',
    hpp: 'text/x-c++',
    cs: 'text/x-csharp',
    php: 'text/x-php',
    dart: 'text/x-dart',
    lua: 'text/x-lua',
    r: 'text/x-r',
    scala: 'text/x-scala',
    ini: 'text/plain',
    cfg: 'text/plain',
    conf: 'text/plain',
    log: 'text/plain',
    dockerfile: 'text/x-dockerfile',
    makefile: 'text/x-makefile',
    gitignore: 'text/plain',
};
export const SUPPORTED_ATTACHMENT_MIME_TYPES = [
    ...Object.values(IMAGE_EXTENSION_MIME_MAP),
    ...Object.values(DOCUMENT_EXTENSION_MIME_MAP),
    ...Object.values(CODE_EXTENSION_MIME_MAP),
];
export const SUPPORTED_ATTACHMENT_EXTENSIONS = [
    ...Object.keys(CODE_EXTENSION_MIME_MAP),
    ...Object.keys(IMAGE_EXTENSION_MIME_MAP),
    ...Object.keys(DOCUMENT_EXTENSION_MIME_MAP),
];
export const MAX_ATTACHMENT_SIZE_BYTES = 5 * 1024 * 1024;
export const MAX_ATTACHMENTS_PER_MESSAGE = 10;
