// @mintlify/models/dist/types/creditTier.js
export const CreditTierIds = [
    'tier-5000',
    'tier-15250',
    'tier-31000',
    'tier-57500',
    'tier-85000',
    'tier-113500',
    'tier-280000',
    'tier-560000',
    'tier-855000',
    'tier-1155000',
];
export const CREDIT_TIER_OVERAGE_PRICE_PER_CREDIT = 0.01;
export const CREDIT_TIERS = {
    'tier-5000': {
        name: '5,000',
        credits: 5000,
        monthlyPrice: 0,
    },
    'tier-15250': {
        name: '15,250',
        credits: 15250,
        monthlyPrice: 100,
    },
    'tier-31000': {
        name: '31,000',
        credits: 31000,
        monthlyPrice: 250,
    },
    'tier-57500': {
        name: '57,500',
        credits: 57500,
        monthlyPrice: 500,
    },
    'tier-85000': {
        name: '85,000',
        credits: 85000,
        monthlyPrice: 750,
    },
    'tier-113500': {
        name: '113,500',
        credits: 113500,
        monthlyPrice: 1000,
    },
    'tier-280000': {
        name: '280,000',
        credits: 280000,
        monthlyPrice: 2500,
    },
    'tier-560000': {
        name: '560,000',
        credits: 560000,
        monthlyPrice: 5000,
    },
    'tier-855000': {
        name: '855,000',
        credits: 855000,
        monthlyPrice: 7500,
    },
    'tier-1155000': {
        name: '1,155,000',
        credits: 1155000,
        monthlyPrice: 10000,
    },
};
