// @mintlify/models/dist/types/onboardingTemplate.js
const DEVELOPER_DOCS_ID = 'developer-docs';
const HELP_CENTER_ID = 'help-center';
const KNOWLEDGE_BASE_ID = 'knowledge-base';
const PRODUCT_GUIDE_ID = 'product-guide';
const EMPTY_ID = 'empty';
export const DEFAULT_ONBOARDING_TEMPLATE_ID = DEVELOPER_DOCS_ID;
export const ONBOARDING_TEMPLATES = [
    {
        id: DEVELOPER_DOCS_ID,
        source: { repo: 'https://github.com/mintlify/starter' },
    },
    {
        id: HELP_CENTER_ID,
        source: { repo: 'https://github.com/mintlify/help-center-starter' },
    },
    {
        id: KNOWLEDGE_BASE_ID,
        source: { repo: 'https://github.com/mintlify/knowledge-base-starter' },
    },
    {
        id: PRODUCT_GUIDE_ID,
        source: { repo: 'https://github.com/mintlify/product-guide-starter' },
    },
    {
        id: EMPTY_ID,
        source: { repo: 'https://github.com/mintlify/empty-starter' },
    },
];
const idSet = new Set(ONBOARDING_TEMPLATES.map((t) => t.id));
export function isOnboardingTemplateIdString(id) {
    return idSet.has(id);
}
export function getOnboardingTemplateById(id) {
    for (const entry of ONBOARDING_TEMPLATES) {
        if (entry.id === id)
            return entry;
    }
    throw new Error(`Unhandled onboarding template id: ${id}`);
}
/**
 * Resolves an optional onboarding template id to a known {@link OnboardingTemplateId}.
 *
 * Returns the default when `id` is `null`, `undefined`, or empty. Throws if a non-empty
 * value is not a registered template id, so callers cannot accidentally pass arbitrary
 * (e.g. user-supplied) strings through to {@link getOnboardingTemplateById}.
 */
export function defaultOnboardingTemplateIdIfOmitted(id) {
    if (id == null || id === '') {
        return DEFAULT_ONBOARDING_TEMPLATE_ID;
    }
    if (!isOnboardingTemplateIdString(id)) {
        throw new Error(`Unknown onboarding template id: ${id}`);
    }
    return id;
}
