// @mintlify/common/dist/mdx/plugins/remark/remarkMermaid.js
import { visit } from 'unist-util-visit';
const REACT_COMPONENT_NAME = 'Mermaid';
export const remarkMermaid = () => (tree) => {
    const codeblocks = [];
    visit(tree, 'code', (node, index, parent) => {
        if (node.lang === 'mermaid' && parent && typeof index === 'number') {
            codeblocks.push([node, index, parent]);
        }
    });
    if (codeblocks.length !== 0) {
        for (const [node, index, parent] of codeblocks) {
            parent.children.splice(index, 1, {
                type: 'mdxJsxFlowElement',
                name: REACT_COMPONENT_NAME,
                attributes: [
                    {
                        type: 'mdxJsxAttribute',
                        name: 'chart',
                        value: {
                            type: 'mdxJsxAttributeValueExpression',
                            data: {
                                estree: {
                                    body: [
                                        {
                                            type: 'ExpressionStatement',
                                            expression: {
                                                type: 'Literal',
                                                value: node.value,
                                                raw: JSON.stringify(node.value),
                                            },
                                        },
                                    ],
                                    type: 'Program',
                                    sourceType: 'module',
                                },
                            },
                            value: node.value,
                        },
                    },
                ],
                children: [],
            });
        }
    }
};
