// @mintlify/common/dist/asyncapi/parser/extractSchemaProperties.js
import { getEnumValues } from '../getEnumValues.js';
const getSchemaType = (value, keys) => {
    if ('type' in value) {
        return value.type;
    }
    const customTypeKey = keys.find((key) => /^x-.*type$/.test(key));
    if (customTypeKey) {
        return value[customTypeKey];
    }
    const enumValues = getEnumValues(value);
    if (enumValues === null || enumValues === void 0 ? void 0 : enumValues.length) {
        const firstValue = enumValues[0];
        if (typeof firstValue === 'boolean')
            return 'boolean';
        if (typeof firstValue === 'number')
            return 'number';
        return 'string';
    }
    return undefined;
};
export const extractSchemaProperties = (schema) => {
    if (!schema) {
        return [];
    }
    const properties = [];
    const requiredFields = new Set(schema.required || []);
    for (const [key, value] of Object.entries(schema)) {
        const name = key;
        if (name === 'required' || name.includes('x-parser')) {
            continue;
        }
        if (name === 'oneOf' || name === 'anyOf' || name === 'allOf') {
            const descriptions = {
                oneOf: 'Must be one of these types',
                anyOf: 'Can be any of these types',
                allOf: 'Must contain all of these types',
            };
            const prop = {
                name,
                type: name,
                description: descriptions[name],
                properties: value
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    .map((item) => {
                    const schemaWithRequired = Object.assign(Object.assign({}, (item.properties || item)), { required: item.required });
                    const extracted = extractSchemaProperties(schemaWithRequired);
                    return extracted;
                })
                    .flat(),
            };
            properties.push(prop);
            continue;
        }
        if (value === null || value === undefined) {
            continue;
        }
        if (typeof value === 'object' && !Array.isArray(value)) {
            const keys = Object.keys(value);
            const type = getSchemaType(value, keys);
            if (type !== undefined) {
                const prop = {
                    name: name,
                    type,
                    description: value.description || value['const'],
                    enumValues: getEnumValues(value),
                    deprecated: value.deprecated,
                    required: requiredFields.has(name),
                };
                if (value.properties) {
                    prop.properties = extractSchemaProperties(value.properties);
                }
                properties.push(prop);
            }
            else {
                properties.push(...extractSchemaProperties(value));
            }
        }
        else {
            properties.push({
                name,
                type: Array.isArray(value) ? 'array' : typeof value,
                description: typeof value === 'string' ? value : undefined,
                required: requiredFields.has(name),
            });
        }
    }
    return properties;
};
