// @mintlify/validation/dist/mint-config/schemas/v2/properties/api.js
import { z } from 'zod';
import { asyncApiSchema } from './reusable/asyncapi.js';
import { openApiSchema } from './reusable/openapi.js';
export const apiSchema = z
    .object({
    openapi: openApiSchema.optional(),
    asyncapi: asyncApiSchema.optional(),
    params: z
        .object({
        expanded: z
            .enum(['all', 'closed'])
            .optional()
            .describe('The view mode of the API parameters. Defaults to `closed`.'),
        post: z
            .array(z.string())
            .optional()
            .describe('OpenAPI spec field keys to render as post pills on every ParamField. For each key, the field value on the schema is surfaced as a pill (string → verbatim, boolean true → key name, array → one pill per element, object → skipped).'),
    })
        .optional()
        .describe('Configurations for the API parameters'),
    playground: z
        .object({
        display: z
            .enum(['interactive', 'simple', 'none', 'auth'])
            .optional()
            .describe('The display mode of the API playground. Defaults to `interactive`. Use `auth` to show the playground only to authenticated users.'),
        proxy: z
            .boolean()
            .optional()
            .describe('Whether to pass API requests through a proxy server. Defaults to `true`.'),
        credentials: z
            .boolean()
            .optional()
            .describe('Whether to include cookies/auth for cross-origin requests when proxy is disabled. Defaults to `false`.'),
    })
        .optional()
        .describe('Configurations for the API playground'),
    examples: z
        .object({
        defaults: z
            .enum(['required', 'all'])
            .optional()
            .describe('Whether to show only required parameters in the examples, defaults to `all`.'),
        languages: z
            .array(z.string())
            .optional()
            .describe('Example languages for the autogenerated API snippets'),
        prefill: z
            .boolean()
            .optional()
            .describe('Whether to prefill the API playground with the examples'),
        autogenerate: z
            .boolean()
            .optional()
            .describe('Whether to generate code samples for API endpoints. Defaults to `true`. If `false`, only manually-written code samples (from either `x-codeSamples` property in the OpenAPI spec or `<RequestExample>` components in the MDX files) will be shown.'),
    })
        .optional()
        .describe('Configurations for the autogenerated API examples'),
    url: z
        .enum(['full'])
        .optional()
        .describe('The display mode of the base URL in the endpoint header. Set to `full` to always show the full base URL on every endpoint page.'),
    mdx: z
        .object({
        auth: z
            .object({
            method: z
                .enum(['bearer', 'basic', 'key', 'cobo'])
                .optional()
                .describe('Authentication method for the API'),
            name: z.string().optional().describe('Authentication name for the API'),
        })
            .optional()
            .describe('Authentication configuration for the API'),
        server: z
            .union([z.string().url(), z.array(z.string().url())])
            .optional()
            .describe('Base URL(s) for the API'),
    })
        .optional()
        .describe('Configurations for API pages generated from MDX files'),
})
    .describe('API reference configuration and playground settings');
