// @mintlify/validation/dist/mint-config/schemas/v2/properties/contextual.js
import { z } from 'zod';
import { iconSchema } from './reusable/icon.js';
export const contextualOptions = [
    'assistant',
    'copy',
    'view',
    'chatgpt',
    'claude',
    'perplexity',
    'grok',
    'aistudio',
    'devin',
    'windsurf',
    'mcp',
    'add-mcp',
    'cursor',
    'vscode',
    'devin-mcp',
];
export const contextualDisplayOptions = ['header', 'toc'];
export const contextualHrefPresetValues = ['$page', '$path', '$mcp'];
export const DetailedContextualHrefSchema = z.object({
    base: z.string(),
    query: z
        .array(z.object({
        key: z.string(),
        value: z.string(),
    }))
        .optional(),
});
export const customContextualOption = z
    .object({
    title: z.string(),
    description: z.string(),
    icon: iconSchema.optional(),
    href: z.union([z.string(), DetailedContextualHrefSchema]),
})
    .describe('Custom contextual option');
export const contextualSchema = z
    .object({
    options: z
        .array(z.union([z.enum(contextualOptions), customContextualOption]))
        .describe('Contextual options')
        .refine((options) => {
        const uniqueOptions = new Set(options);
        return uniqueOptions.size === options.length;
    }, {
        message: 'Each contextual option can only appear once',
    }),
    display: z
        .enum(contextualDisplayOptions)
        .optional()
        .default('header')
        .describe('Where to display the contextual options: "header" (default) shows them in the top-of-page context menu, "toc" shows them in the table of contents sidebar'),
})
    .describe('Contextual options');
