// @mintlify/validation/dist/mint-config/schemas/v2/properties/navbar.js
import { z } from 'zod';
import { hrefSchema } from './reusable/href.js';
import { iconSchema } from './reusable/icon.js';
const baseLinkSchema = z.object({
    icon: iconSchema.optional(),
    href: hrefSchema,
});
const standardLinkSchema = baseLinkSchema.extend({
    label: z.string(),
    type: z.string().optional(),
});
const githubLinkSchema = baseLinkSchema.extend({
    label: z.string().optional(),
    type: z.literal('github'),
});
const discordLinkSchema = baseLinkSchema.extend({
    label: z.string().optional(),
    type: z.literal('discord'),
});
const linkSchema = z.union([githubLinkSchema, discordLinkSchema, standardLinkSchema]);
const buttonSchema = z.object({
    type: z.literal('button'),
    label: z.string(),
    href: hrefSchema,
});
const githubSchema = z.object({
    type: z.literal('github'),
    label: z.string().optional(),
    href: hrefSchema,
});
const discordSchema = z.object({
    type: z.literal('discord'),
    label: z.string().optional(),
    href: hrefSchema,
});
export const primarySchema = z.discriminatedUnion('type', [
    buttonSchema,
    githubSchema,
    discordSchema,
]);
export const navbarSchema = z
    .object({
    links: z.array(linkSchema).optional().describe('The links in the navbar'),
    primary: primarySchema.optional().describe('The primary CTA in the navbar'),
})
    .describe('Navbar content and settings');
