// @mintlify/validation/dist/mint-config/schemas/v2/properties/navigation/groups.js
import { z } from 'zod';
import { asyncApiSchema } from '../reusable/asyncapi.js';
import { hiddenSchema } from '../reusable/hidden.js';
import { iconSchema } from '../reusable/icon.js';
import { openApiSchema } from '../reusable/openapi.js';
import { decoratedPageSchema, pageSchema } from '../reusable/page.js';
import { decoratedPagesSchema, pagesSchema } from './pages.js';
/** How automatic directory UI is shown for group roots under this navigation node. Inherits through `navigation`. */
export const NAVIGATION_DIRECTORY = ['none', 'accordion', 'card'];
export const directoryField = {
    directory: z
        .enum(NAVIGATION_DIRECTORY)
        .optional()
        .describe('Directory layout for this navigation node and its descendants. "accordion" and "card" show automatic directory UI on eligible root pages; "none" disables it. Inherits recursively; descendants can override.'),
};
export const baseGroupSchema = z.object(Object.assign({ group: z.string().nonempty().describe('The name of the group'), public: z.boolean().optional().describe('Whether the group is public'), boost: z
        .number()
        .optional()
        .describe('Search ranking multiplier for pages in this group (e.g. 3 = 3x boost)'), icon: iconSchema.optional(), hidden: hiddenSchema.optional(), root: pageSchema.optional(), tag: z.string().optional().describe('Tag for the group'), expanded: z.boolean().optional().describe('Whether the group is expanded by default') }, directoryField));
export const groupSchema = z
    .union([
    baseGroupSchema.extend({
        openapi: openApiSchema,
        pages: z.lazy(() => pagesSchema).default([]),
    }),
    baseGroupSchema.extend({
        asyncapi: asyncApiSchema,
        pages: z.lazy(() => pagesSchema).default([]),
    }),
    baseGroupSchema.extend({ pages: z.lazy(() => pagesSchema).default([]) }),
])
    .describe('Organizing by groups');
export const decoratedGroupSchema = z
    .object(Object.assign({ group: z.string().nonempty().describe('The name of the group'), public: z.boolean().optional().describe('Whether the group is public'), boost: z
        .number()
        .optional()
        .describe('Search ranking multiplier for pages in this group (e.g. 3 = 3x boost)'), icon: iconSchema.optional(), hidden: hiddenSchema.optional(), root: decoratedPageSchema.optional(), pages: z.lazy(() => decoratedPagesSchema), tag: z.string().optional().describe('Tag for the group'), expanded: z.boolean().optional().describe('Whether the group is expanded by default') }, directoryField))
    .describe('Organizing by groups');
export const groupsSchema = z
    .array(groupSchema)
    .min(1, 'At least one group must be specified')
    .describe('Organizing by groups');
export const decoratedGroupsSchema = z
    .array(decoratedGroupSchema)
    .min(1, 'At least one group must be specified')
    .describe('Organizing by groups');
