// @mintlify/prebuild/dist/prebuild/update/mintConfig/index.js
import { getConfigPath } from '../../../utils.js';
import { MintConfigUpdater } from '../ConfigUpdater.js';
import { generateOpenApiAnchorsOrTabs } from './generateOpenApiAnchorsOrTabs.js';
export async function updateMintConfigFile(contentDirectoryPath, openApiFiles, localSchema, strict) {
    const configPath = await getConfigPath(contentDirectoryPath, 'mint');
    if (configPath == null) {
        return null;
    }
    const mintConfig = await MintConfigUpdater.getConfig(configPath, strict, contentDirectoryPath);
    const { mintConfig: newMintConfig, pagesAcc, openApiFiles: newOpenApiFiles, } = await generateOpenApiAnchorsOrTabs(mintConfig, openApiFiles, undefined, localSchema);
    await MintConfigUpdater.writeConfigFile(newMintConfig);
    return {
        mintConfig: newMintConfig,
        pagesAcc,
        openApiFiles: newOpenApiFiles,
        originalMintConfig: mintConfig,
    };
}
export { generateOpenApiAnchorsOrTabs, generateOpenApiAnchorOrTab, } from './generateOpenApiAnchorsOrTabs.js';
