// @mintlify/validation/dist/editor-navigation/conversion/common.js
import { generateNodeId } from '../utils.js';
export const createIdGenerator = () => {
    let counter = 0;
    return () => generateNodeId(counter++);
};
export function isProductConfig(value) {
    return typeof value === 'object' && value !== null && 'product' in value;
}
export function isLanguageConfig(value) {
    return typeof value === 'object' && value !== null && 'language' in value;
}
export function isVersionConfig(value) {
    return typeof value === 'object' && value !== null && 'version' in value;
}
export function isTabConfig(value) {
    return typeof value === 'object' && value !== null && 'tab' in value;
}
export function isMenuItemConfig(value) {
    return typeof value === 'object' && value !== null && 'item' in value;
}
export function isDropdownConfig(value) {
    return typeof value === 'object' && value !== null && 'dropdown' in value;
}
export function isAnchorConfig(value) {
    return typeof value === 'object' && value !== null && 'anchor' in value;
}
export function isGroupConfig(value) {
    return typeof value === 'object' && value !== null && 'group' in value;
}
export function buildChildrenMap(tree) {
    const childrenMap = new Map();
    tree.forEach((node) => {
        const children = childrenMap.get(node.parent) || [];
        children.push(node);
        childrenMap.set(node.parent, children);
    });
    childrenMap.forEach((children) => {
        children.sort((a, b) => a.order - b.order);
    });
    return childrenMap;
}
export function mapNodesOfType(nodes, type, childrenMap, mapper) {
    return nodes.filter((n) => n.type === type).map((node) => mapper(node, childrenMap));
}
export function assignIfPresent(target, source, keys) {
    for (const key of keys) {
        const value = source[key];
        if (value !== undefined) {
            target[key] = value;
        }
    }
}
export const COMMON_DIVISION_NODE_DATA_FIELDS = [
    'hidden',
    'href',
    'asyncapi',
    'openapi',
    'global',
];
export const TAB_NODE_DATA_FIELDS = ['icon', ...COMMON_DIVISION_NODE_DATA_FIELDS];
export const ANCHOR_NODE_DATA_FIELDS = [
    'icon',
    'color',
    ...COMMON_DIVISION_NODE_DATA_FIELDS,
];
export const LANGUAGE_NODE_DATA_FIELDS = ['default', ...COMMON_DIVISION_NODE_DATA_FIELDS];
export const VERSION_NODE_DATA_FIELDS = [
    'default',
    'tag',
    ...COMMON_DIVISION_NODE_DATA_FIELDS,
];
export const DROPDOWN_NODE_DATA_FIELDS = [
    'icon',
    'color',
    'description',
    ...COMMON_DIVISION_NODE_DATA_FIELDS,
];
export const PRODUCT_NODE_DATA_FIELDS = [
    'name',
    'icon',
    'color',
    'description',
    ...COMMON_DIVISION_NODE_DATA_FIELDS,
];
export const MENU_ITEM_NODE_DATA_FIELDS = [
    'icon',
    'description',
    ...COMMON_DIVISION_NODE_DATA_FIELDS,
];
export const GROUP_NODE_DATA_FIELDS = [
    'icon',
    'hidden',
    'tag',
    'expanded',
    'public',
    'root',
    'asyncapi',
    'openapi',
];
