// @mintlify/validation/dist/editor-navigation/conversion/docsjson-to-tree.js
import traverse from 'neotraverse';
import { isGlobalNode } from '../schema.js';
import { createIdGenerator, isProductConfig, isLanguageConfig, isVersionConfig, isTabConfig, isMenuItemConfig, isDropdownConfig, isAnchorConfig, isGroupConfig, } from './common.js';
function pickDefined(obj) {
    const result = {};
    for (const key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key)) {
            const value = obj[key];
            if (value !== undefined && value !== null) {
                result[key] = value;
            }
        }
    }
    return result;
}
const NODE_FACTORIES = {
    products: ({ value, parent, generateId, getNextOrder }) => {
        if (!isProductConfig(value))
            return null;
        const nodeId = generateId();
        const productNode = {
            id: nodeId,
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.name || value.product,
            type: 'product',
            isHidden: value.hidden === true,
            data: Object.assign({ product: value.product }, pickDefined({
                name: value.name,
                icon: value.icon,
                color: value.color,
                description: value.description,
                hidden: value.hidden,
                openapi: 'openapi' in value ? value.openapi : undefined,
                asyncapi: 'asyncapi' in value ? value.asyncapi : undefined,
            })),
        };
        return productNode;
    },
    languages: ({ value, parent, generateId, getNextOrder }) => {
        if (!isLanguageConfig(value))
            return null;
        const nodeId = generateId();
        const langNode = {
            id: nodeId,
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.language,
            type: 'language',
            isHidden: value.hidden === true,
            data: Object.assign({ language: value.language }, pickDefined({
                default: value.default,
                hidden: value.hidden,
                openapi: 'openapi' in value ? value.openapi : undefined,
                asyncapi: 'asyncapi' in value ? value.asyncapi : undefined,
            })),
        };
        return langNode;
    },
    versions: ({ value, parent, generateId, getNextOrder }) => {
        if (!isVersionConfig(value))
            return null;
        const nodeId = generateId();
        const verNode = {
            id: nodeId,
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.version,
            type: 'version',
            isHidden: value.hidden === true,
            data: Object.assign({ version: value.version }, pickDefined({
                default: value.default,
                hidden: value.hidden,
                tag: value.tag,
                openapi: 'openapi' in value ? value.openapi : undefined,
                asyncapi: 'asyncapi' in value ? value.asyncapi : undefined,
            })),
        };
        return verNode;
    },
    tabs: ({ value, parent, generateId, getNextOrder }) => {
        if (!isTabConfig(value))
            return null;
        const nodeId = generateId();
        const tabNode = {
            id: nodeId,
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.tab,
            type: 'tab',
            isHidden: value.hidden === true,
            data: Object.assign({ tab: value.tab }, pickDefined({
                icon: value.icon,
                hidden: value.hidden,
                href: 'href' in value ? value.href : undefined,
                openapi: 'openapi' in value ? value.openapi : undefined,
                asyncapi: 'asyncapi' in value ? value.asyncapi : undefined,
            })),
        };
        return tabNode;
    },
    dropdowns: ({ value, parent, generateId, getNextOrder }) => {
        if (!isDropdownConfig(value))
            return null;
        const nodeId = generateId();
        const dropdownNode = {
            id: nodeId,
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.dropdown,
            type: 'dropdown',
            isHidden: value.hidden === true,
            data: Object.assign({ dropdown: value.dropdown }, pickDefined({
                icon: value.icon,
                color: value.color,
                description: value.description,
                hidden: value.hidden,
                openapi: 'openapi' in value ? value.openapi : undefined,
                asyncapi: 'asyncapi' in value ? value.asyncapi : undefined,
            })),
        };
        return dropdownNode;
    },
    anchors: ({ value, parent, generateId, getNextOrder }) => {
        if (!isAnchorConfig(value))
            return null;
        const nodeId = generateId();
        const anchorNode = {
            id: nodeId,
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.anchor,
            type: 'anchor',
            isHidden: value.hidden === true,
            data: Object.assign({ anchor: value.anchor }, pickDefined({
                icon: value.icon,
                color: value.color,
                hidden: value.hidden,
                href: 'href' in value ? value.href : undefined,
                openapi: 'openapi' in value ? value.openapi : undefined,
                asyncapi: 'asyncapi' in value ? value.asyncapi : undefined,
            })),
        };
        return anchorNode;
    },
    menu: ({ value, parent, generateId, getNextOrder }) => {
        if (!isMenuItemConfig(value))
            return null;
        const nodeId = generateId();
        const menuItemNode = {
            id: nodeId,
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.item,
            type: 'item',
            isHidden: value.hidden === true,
            data: Object.assign({ item: value.item }, pickDefined({
                icon: value.icon,
                description: value.description,
                hidden: value.hidden,
                href: 'href' in value ? value.href : undefined,
                openapi: 'openapi' in value ? value.openapi : undefined,
                asyncapi: 'asyncapi' in value ? value.asyncapi : undefined,
            })),
        };
        return menuItemNode;
    },
    groups: ({ value, parent, generateId, getNextOrder }) => {
        if (!isGroupConfig(value))
            return null;
        const nodeId = generateId();
        const groupNode = {
            id: nodeId,
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.group,
            type: 'group',
            isHidden: value.hidden === true,
            data: Object.assign({ group: value.group }, pickDefined({
                icon: value.icon,
                hidden: value.hidden,
                tag: value.tag,
                expanded: value.expanded,
                public: value.public,
                root: value.root,
                openapi: 'openapi' in value ? value.openapi : undefined,
                asyncapi: 'asyncapi' in value ? value.asyncapi : undefined,
            })),
        };
        return groupNode;
    },
};
const GLOBAL_NODE_FACTORIES = {
    languages: ({ value, parent, generateId, getNextOrder }) => {
        if (!isLanguageConfig(value))
            return null;
        return {
            id: generateId(),
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.language,
            type: 'language',
            isHidden: value.hidden === true,
            data: Object.assign({ language: value.language, isGlobal: true }, pickDefined({
                default: value.default,
                hidden: value.hidden,
                href: 'href' in value ? value.href : undefined,
            })),
        };
    },
    versions: ({ value, parent, generateId, getNextOrder }) => {
        if (!isVersionConfig(value))
            return null;
        return {
            id: generateId(),
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.version,
            type: 'version',
            isHidden: value.hidden === true,
            data: Object.assign({ version: value.version, isGlobal: true }, pickDefined({
                default: value.default,
                hidden: value.hidden,
                tag: value.tag,
                href: 'href' in value ? value.href : undefined,
            })),
        };
    },
    tabs: ({ value, parent, generateId, getNextOrder }) => {
        if (!isTabConfig(value))
            return null;
        return {
            id: generateId(),
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.tab,
            type: 'tab',
            isHidden: value.hidden === true,
            data: Object.assign({ tab: value.tab, isGlobal: true }, pickDefined({
                icon: value.icon,
                hidden: value.hidden,
                href: 'href' in value ? value.href : undefined,
            })),
        };
    },
    dropdowns: ({ value, parent, generateId, getNextOrder }) => {
        if (!isDropdownConfig(value))
            return null;
        return {
            id: generateId(),
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.dropdown,
            type: 'dropdown',
            isHidden: value.hidden === true,
            data: Object.assign({ dropdown: value.dropdown, isGlobal: true }, pickDefined({
                icon: value.icon,
                color: value.color,
                description: value.description,
                hidden: value.hidden,
                href: 'href' in value ? value.href : undefined,
            })),
        };
    },
    anchors: ({ value, parent, generateId, getNextOrder }) => {
        if (!isAnchorConfig(value))
            return null;
        return {
            id: generateId(),
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.anchor,
            type: 'anchor',
            isHidden: value.hidden === true,
            data: Object.assign({ anchor: value.anchor, isGlobal: true }, pickDefined({
                icon: value.icon,
                color: value.color,
                hidden: value.hidden,
                href: 'href' in value ? value.href : undefined,
            })),
        };
    },
    products: ({ value, parent, generateId, getNextOrder }) => {
        if (!isProductConfig(value))
            return null;
        return {
            id: generateId(),
            parent: parent.id,
            order: getNextOrder(parent.id),
            text: value.name || value.product,
            type: 'product',
            isHidden: value.hidden === true,
            data: Object.assign({ product: value.product, isGlobal: true }, pickDefined({
                name: value.name,
                icon: value.icon,
                color: value.color,
                description: value.description,
                hidden: value.hidden,
                href: 'href' in value ? value.href : undefined,
            })),
        };
    },
};
export function docsConfigToNavigationTree(navigation) {
    const nodes = [];
    const generateId = createIdGenerator();
    const parentStack = [{ id: 0, order: 0 }];
    const orderByParent = new Map();
    const getNextOrder = (parentId) => {
        var _a;
        const current = (_a = orderByParent.get(parentId)) !== null && _a !== void 0 ? _a : 0;
        orderByParent.set(parentId, current + 1);
        return current;
    };
    const getCurrentParent = () => {
        var _a;
        return (_a = parentStack[parentStack.length - 1]) !== null && _a !== void 0 ? _a : { id: 0, order: 0 };
    };
    traverse(navigation).forEach(function (value) {
        var _a, _b, _c;
        const parentKey = (_a = this.parent) === null || _a === void 0 ? void 0 : _a.key;
        const grandparentKey = (_c = (_b = this.parent) === null || _b === void 0 ? void 0 : _b.parent) === null || _c === void 0 ? void 0 : _c.key;
        const currentParent = getCurrentParent();
        if (!parentKey || typeof parentKey !== 'string')
            return;
        if (grandparentKey === 'global' && typeof parentKey === 'string') {
            const globalFactory = GLOBAL_NODE_FACTORIES[parentKey];
            if (globalFactory) {
                const node = globalFactory({
                    value,
                    parent: currentParent,
                    generateId,
                    getNextOrder,
                });
                if (node) {
                    nodes.push(node);
                }
            }
            return;
        }
        if (parentKey === 'pages') {
            if (typeof value === 'string') {
                const nodeId = generateId();
                const text = value.split('/').pop() || value;
                const node = {
                    id: nodeId,
                    parent: currentParent.id,
                    order: getNextOrder(currentParent.id),
                    text,
                    type: 'page',
                    data: { title: text, href: value },
                };
                nodes.push(node);
            }
            else if (isGroupConfig(value)) {
                const nodeId = generateId();
                const groupNode = {
                    id: nodeId,
                    parent: currentParent.id,
                    order: getNextOrder(currentParent.id),
                    text: value.group,
                    type: 'group',
                    isHidden: value.hidden === true,
                    data: Object.assign({ group: value.group, fromPages: true }, pickDefined({
                        icon: value.icon,
                        hidden: value.hidden,
                        tag: value.tag,
                        expanded: value.expanded,
                        public: value.public,
                        root: value.root,
                        openapi: 'openapi' in value ? value.openapi : undefined,
                        asyncapi: 'asyncapi' in value ? value.asyncapi : undefined,
                    })),
                };
                nodes.push(groupNode);
                parentStack.push({ id: nodeId, order: 0 });
                this.after(() => parentStack.pop());
            }
            return;
        }
        const factory = NODE_FACTORIES[parentKey];
        if (!factory)
            return;
        const node = factory({
            value,
            parent: currentParent,
            generateId,
            getNextOrder,
        });
        if (!node)
            return;
        nodes.push(node);
        parentStack.push({ id: node.id, order: 0 });
        this.after(() => parentStack.pop());
    });
    reorderGlobalNodesFirst(nodes);
    return nodes;
}
function reorderGlobalNodesFirst(nodes) {
    var _a;
    const nodesByParent = new Map();
    for (const node of nodes) {
        const siblings = (_a = nodesByParent.get(node.parent)) !== null && _a !== void 0 ? _a : [];
        siblings.push(node);
        nodesByParent.set(node.parent, siblings);
    }
    for (const siblings of nodesByParent.values()) {
        siblings.sort((a, b) => {
            const aIsGlobal = isGlobalNode(a);
            const bIsGlobal = isGlobalNode(b);
            if (aIsGlobal && !bIsGlobal)
                return -1;
            if (!aIsGlobal && bIsGlobal)
                return 1;
            return a.order - b.order;
        });
        siblings.forEach((node, index) => {
            node.order = index;
        });
    }
}
