// @mintlify/validation/dist/editor-navigation/conversion/tree-to-docsjson.js
import { isGlobalNode } from '../schema.js';
import { buildChildrenMap, mapNodesOfType, assignIfPresent, TAB_NODE_DATA_FIELDS, ANCHOR_NODE_DATA_FIELDS, LANGUAGE_NODE_DATA_FIELDS, VERSION_NODE_DATA_FIELDS, DROPDOWN_NODE_DATA_FIELDS, PRODUCT_NODE_DATA_FIELDS, GROUP_NODE_DATA_FIELDS, MENU_ITEM_NODE_DATA_FIELDS, } from './common.js';
function getGroupPagesChildren(node, childrenMap) {
    const children = childrenMap.get(node.id) || [];
    if (node.type !== 'group')
        return children;
    const rootHref = node.data.root;
    if (!rootHref)
        return children;
    return children.filter((child) => !(child.type === 'page' && child.data.href === rootHref));
}
function nodesToPages(nodes, childrenMap) {
    return nodes.map((node) => {
        if (node.type === 'group') {
            const groupChildren = getGroupPagesChildren(node, childrenMap);
            const baseConfig = { group: node.data.group };
            assignIfPresent(baseConfig, node.data, GROUP_NODE_DATA_FIELDS);
            if ((node.data.openapi !== undefined || node.data.asyncapi !== undefined) &&
                !groupChildren.length) {
                // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- groups default `pages` to an empty array, so we can cast with pages being undefined
                return Object.assign(Object.assign({}, baseConfig), { group: node.data.group });
            }
            return Object.assign(Object.assign({}, baseConfig), { group: node.data.group, pages: nodesToPages(groupChildren, childrenMap) });
        }
        if (node.type === 'page') {
            return node.data.href;
        }
        return '';
    });
}
function nodesToGroups(nodes, childrenMap) {
    return mapNodesOfType(nodes, 'group', childrenMap, (node, map) => {
        const groupChildren = getGroupPagesChildren(node, map);
        const baseConfig = { group: node.data.group };
        assignIfPresent(baseConfig, node.data, GROUP_NODE_DATA_FIELDS);
        if ((node.data.openapi !== undefined || node.data.asyncapi !== undefined) &&
            !groupChildren.length) {
            // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- groups default `pages` to an empty array, so we can cast with pages being undefined
            return Object.assign(Object.assign({}, baseConfig), { group: node.data.group });
        }
        return Object.assign(Object.assign({}, baseConfig), { group: node.data.group, pages: nodesToPages(groupChildren, map) });
    });
}
function getGlobalChildren(childrenMap, nodeId) {
    const children = childrenMap.get(nodeId) || [];
    return children.filter(isGlobalNode);
}
function getRegularChildren(childrenMap, nodeId) {
    const children = childrenMap.get(nodeId) || [];
    return children.filter((child) => !isGlobalNode(child));
}
function nodesToGlobalConfig(globalChildren) {
    if (globalChildren.length === 0)
        return undefined;
    const globalConfig = {};
    const languages = globalChildren.filter((n) => n.type === 'language');
    const versions = globalChildren.filter((n) => n.type === 'version');
    const tabs = globalChildren.filter((n) => n.type === 'tab');
    const dropdowns = globalChildren.filter((n) => n.type === 'dropdown');
    const anchors = globalChildren.filter((n) => n.type === 'anchor');
    const products = globalChildren.filter((n) => n.type === 'product');
    if (languages.length > 0) {
        globalConfig.languages = languages.map((node) => (Object.assign(Object.assign(Object.assign({ language: node.data.language }, (node.data.hidden !== undefined && { hidden: node.data.hidden })), (node.data.default !== undefined && { default: node.data.default })), { href: node.data.href || '' })));
    }
    if (versions.length > 0) {
        globalConfig.versions = versions.map((node) => (Object.assign(Object.assign(Object.assign(Object.assign({ version: node.data.version }, (node.data.hidden !== undefined && { hidden: node.data.hidden })), (node.data.default !== undefined && { default: node.data.default })), (node.data.tag !== undefined && { tag: node.data.tag })), { href: node.data.href || '' })));
    }
    if (tabs.length > 0) {
        globalConfig.tabs = tabs.map((node) => (Object.assign(Object.assign(Object.assign({ tab: node.data.tab }, (node.data.icon !== undefined && { icon: node.data.icon })), (node.data.hidden !== undefined && { hidden: node.data.hidden })), { href: node.data.href || '' })));
    }
    if (dropdowns.length > 0) {
        globalConfig.dropdowns = dropdowns.map((node) => (Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({ dropdown: node.data.dropdown }, (node.data.icon !== undefined && { icon: node.data.icon })), (node.data.color !== undefined && { color: node.data.color })), (node.data.description !== undefined && { description: node.data.description })), (node.data.hidden !== undefined && { hidden: node.data.hidden })), { href: node.data.href || '' })));
    }
    if (anchors.length > 0) {
        globalConfig.anchors = anchors.map((node) => (Object.assign(Object.assign(Object.assign(Object.assign({ anchor: node.data.anchor }, (node.data.icon !== undefined && { icon: node.data.icon })), (node.data.color !== undefined && { color: node.data.color })), (node.data.hidden !== undefined && { hidden: node.data.hidden })), { href: node.data.href || '' })));
    }
    if (products.length > 0) {
        globalConfig.products = products.map((node) => (Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({ product: node.data.product }, (node.data.name !== undefined && { name: node.data.name })), (node.data.icon !== undefined && { icon: node.data.icon })), (node.data.color !== undefined && { color: node.data.color })), (node.data.description !== undefined && { description: node.data.description })), (node.data.hidden !== undefined && { hidden: node.data.hidden })), { href: node.data.href || '' })));
    }
    return Object.keys(globalConfig).length > 0 ? globalConfig : undefined;
}
function buildChildNavigation(node, childrenMap) {
    const regularChildren = getRegularChildren(childrenMap, node.id);
    if (regularChildren.length === 0) {
        return {};
    }
    const firstChild = regularChildren[0];
    if (!firstChild) {
        return {};
    }
    const childType = firstChild.type;
    const children = regularChildren;
    switch (childType) {
        case 'product':
            return { products: nodesToProducts(children, childrenMap) };
        case 'language':
            return { languages: nodesToLanguages(children, childrenMap) };
        case 'version':
            return { versions: nodesToVersions(children, childrenMap) };
        case 'tab':
            return { tabs: nodesToTabs(children, childrenMap) };
        case 'dropdown':
            return { dropdowns: nodesToDropdowns(children, childrenMap) };
        case 'anchor':
            return { anchors: nodesToAnchors(children, childrenMap) };
        case 'item':
            return { menu: nodesToMenuItems(children, childrenMap) };
        case 'group':
            if (firstChild.data.fromPages) {
                return { pages: nodesToPages(children, childrenMap) };
            }
            const hasPageSibling = children.some((child) => child.type === 'page');
            if (hasPageSibling) {
                return { pages: nodesToPages(children, childrenMap) };
            }
            return { groups: nodesToGroups(children, childrenMap) };
        case 'page':
            return { pages: nodesToPages(children, childrenMap) };
        default:
            return {};
    }
}
function nodesToTabs(nodes, childrenMap) {
    return mapNodesOfType(nodes, 'tab', childrenMap, (node, map) => {
        const globalChildren = getGlobalChildren(map, node.id);
        const globalConfig = nodesToGlobalConfig(globalChildren);
        const baseConfig = { tab: node.data.tab };
        assignIfPresent(baseConfig, node.data, TAB_NODE_DATA_FIELDS);
        if (node.data.href && !node.data.isGlobal) {
            return Object.assign(Object.assign(Object.assign({}, baseConfig), { tab: node.data.tab, href: node.data.href }), (globalConfig && { global: globalConfig }));
        }
        const childNav = buildChildNavigation(node, map);
        return Object.assign(Object.assign(Object.assign(Object.assign({}, baseConfig), { tab: node.data.tab }), (globalConfig && { global: globalConfig })), childNav);
    });
}
function nodesToAnchors(nodes, childrenMap) {
    return mapNodesOfType(nodes, 'anchor', childrenMap, (node, map) => {
        const globalChildren = getGlobalChildren(map, node.id);
        const globalConfig = nodesToGlobalConfig(globalChildren);
        const baseConfig = { anchor: node.data.anchor };
        assignIfPresent(baseConfig, node.data, ANCHOR_NODE_DATA_FIELDS);
        if (node.data.href && !node.data.isGlobal) {
            return Object.assign(Object.assign(Object.assign({}, baseConfig), { anchor: node.data.anchor, href: node.data.href }), (globalConfig && { global: globalConfig }));
        }
        const childNav = buildChildNavigation(node, map);
        return Object.assign(Object.assign(Object.assign(Object.assign({}, baseConfig), { anchor: node.data.anchor }), (globalConfig && { global: globalConfig })), childNav);
    });
}
function nodesToLanguages(nodes, childrenMap) {
    return mapNodesOfType(nodes, 'language', childrenMap, (node, map) => {
        const globalChildren = getGlobalChildren(map, node.id);
        const globalConfig = nodesToGlobalConfig(globalChildren);
        const baseConfig = Object.assign({ language: node.data.language }, (node.data.default !== undefined && { default: node.data.default }));
        assignIfPresent(baseConfig, node.data, LANGUAGE_NODE_DATA_FIELDS);
        if (node.data.href && !node.data.isGlobal) {
            return Object.assign(Object.assign(Object.assign({}, baseConfig), { language: node.data.language, href: node.data.href }), (globalConfig && { global: globalConfig }));
        }
        const childNav = buildChildNavigation(node, map);
        return Object.assign(Object.assign(Object.assign(Object.assign({}, baseConfig), { language: node.data.language }), (globalConfig && { global: globalConfig })), childNav);
    });
}
function nodesToVersions(nodes, childrenMap) {
    return mapNodesOfType(nodes, 'version', childrenMap, (node, map) => {
        const globalChildren = getGlobalChildren(map, node.id);
        const globalConfig = nodesToGlobalConfig(globalChildren);
        const baseConfig = Object.assign({ version: node.data.version }, (node.data.default !== undefined && { default: node.data.default }));
        assignIfPresent(baseConfig, node.data, VERSION_NODE_DATA_FIELDS);
        if (node.data.href && !node.data.isGlobal) {
            return Object.assign(Object.assign(Object.assign({}, baseConfig), { version: node.data.version, href: node.data.href }), (globalConfig && { global: globalConfig }));
        }
        const childNav = buildChildNavigation(node, map);
        return Object.assign(Object.assign(Object.assign(Object.assign({}, baseConfig), { version: node.data.version }), (globalConfig && { global: globalConfig })), childNav);
    });
}
function nodesToDropdowns(nodes, childrenMap) {
    return mapNodesOfType(nodes, 'dropdown', childrenMap, (node, map) => {
        const globalChildren = getGlobalChildren(map, node.id);
        const globalConfig = nodesToGlobalConfig(globalChildren);
        const baseConfig = { dropdown: node.data.dropdown };
        assignIfPresent(baseConfig, node.data, DROPDOWN_NODE_DATA_FIELDS);
        if (node.data.href && !node.data.isGlobal) {
            return Object.assign(Object.assign(Object.assign({}, baseConfig), { dropdown: node.data.dropdown, href: node.data.href }), (globalConfig && { global: globalConfig }));
        }
        const childNav = buildChildNavigation(node, map);
        return Object.assign(Object.assign(Object.assign(Object.assign({}, baseConfig), { dropdown: node.data.dropdown }), (globalConfig && { global: globalConfig })), childNav);
    });
}
function nodesToProducts(nodes, childrenMap) {
    return mapNodesOfType(nodes, 'product', childrenMap, (node, map) => {
        const globalChildren = getGlobalChildren(map, node.id);
        const globalConfig = nodesToGlobalConfig(globalChildren);
        const baseConfig = { product: node.data.product };
        assignIfPresent(baseConfig, node.data, PRODUCT_NODE_DATA_FIELDS);
        if (node.data.href && !node.data.isGlobal) {
            return Object.assign(Object.assign(Object.assign({}, baseConfig), { product: node.data.product, href: node.data.href }), (globalConfig && { global: globalConfig }));
        }
        const childNav = buildChildNavigation(node, map);
        return Object.assign(Object.assign(Object.assign(Object.assign({}, baseConfig), { product: node.data.product }), (globalConfig && { global: globalConfig })), childNav);
    });
}
function nodesToMenuItems(nodes, childrenMap) {
    return mapNodesOfType(nodes, 'item', childrenMap, (node, map) => {
        const globalChildren = getGlobalChildren(map, node.id);
        const globalConfig = nodesToGlobalConfig(globalChildren);
        const baseConfig = { item: node.data.item };
        assignIfPresent(baseConfig, node.data, MENU_ITEM_NODE_DATA_FIELDS);
        if (node.data.href) {
            return Object.assign(Object.assign(Object.assign({}, baseConfig), { item: node.data.item, href: node.data.href }), (globalConfig && { global: globalConfig }));
        }
        const childNav = buildChildNavigation(node, map);
        return Object.assign(Object.assign(Object.assign(Object.assign({}, baseConfig), { item: node.data.item }), (globalConfig && { global: globalConfig })), childNav);
    });
}
export function navigationTreeToDocsConfig(tree) {
    const childrenMap = buildChildrenMap(tree);
    const rootGlobalChildren = getGlobalChildren(childrenMap, 0);
    const rootRegularChildren = getRegularChildren(childrenMap, 0);
    const globalConfig = nodesToGlobalConfig(rootGlobalChildren);
    if (rootRegularChildren.length === 0) {
        if (globalConfig) {
            return { global: globalConfig, groups: [] };
        }
        return { groups: [] };
    }
    const firstChild = rootRegularChildren[0];
    if (!firstChild) {
        if (globalConfig) {
            return { global: globalConfig, groups: [] };
        }
        return { groups: [] };
    }
    const rootType = firstChild.type;
    let result;
    switch (rootType) {
        case 'product':
            result = { products: nodesToProducts(rootRegularChildren, childrenMap) };
            break;
        case 'language':
            result = { languages: nodesToLanguages(rootRegularChildren, childrenMap) };
            break;
        case 'version':
            result = { versions: nodesToVersions(rootRegularChildren, childrenMap) };
            break;
        case 'tab':
            result = { tabs: nodesToTabs(rootRegularChildren, childrenMap) };
            break;
        case 'dropdown':
            result = { dropdowns: nodesToDropdowns(rootRegularChildren, childrenMap) };
            break;
        case 'anchor':
            result = { anchors: nodesToAnchors(rootRegularChildren, childrenMap) };
            break;
        case 'group': {
            const hasPageSibling = rootRegularChildren.some((child) => child.type === 'page');
            if (hasPageSibling) {
                result = { pages: nodesToPages(rootRegularChildren, childrenMap) };
            }
            else {
                result = { groups: nodesToGroups(rootRegularChildren, childrenMap) };
            }
            break;
        }
        case 'page':
            result = { pages: nodesToPages(rootRegularChildren, childrenMap) };
            break;
        default:
            result = { groups: [] };
    }
    if (globalConfig) {
        return Object.assign({ global: globalConfig }, result);
    }
    return result;
}
